cmake/modules/ccache.cmake

```diff
@@ -3,6 +3,8 @@
 # Use ccache if it is installed, unless the user explicitly disables
 # it by setting USE_CCACHE=0.
 
+include_guard(GLOBAL)
+
```

Comment:
Why make this code into a module, and why do we document the input and output of some modules, but not others?

Reply:
I think this falls in line with this answer: https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r793936366

I'm are not documenting some files based on whether they are modules or not.

I ensure **new** files are properly described in the beginning with their purpose, when adding new files to the repository.

Reply:
and to also answer first part of the question:
> Why make this code into a module

to ensure that all the CMake code that are directly loaded from [zephyr_default.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/6873c4c361a87514d57a1eef29342ec026d4b3de/cmake/modules/zephyr_default.cmake) can be loaded as modules. (only exception is the `pre_dt_board.cmake` which is located in `board` specific folder).
`ccache.cmake` is an existing file called directly from boilerplate, and preference has been to minimize changes to execution flow with this PR.
After the new structure is in place, then additional work can be done to improve the modular approach.

`ccache.cmake` shall probably in future be closer integrated to toolchain handling, but that will be continued work, similar to modules also mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r790009740

Reply:
Why should all code loaded from zephyr_default.cmake be loaded as modules?

Reply:
because else we can end in same situation as with existing boilerplate code, where a user cannot opt-out of a given part, or cannot load only a sub-part of the system without loading everything.

I already presented two Zephyr specific use-cases here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#issue-1083225661
that suffered from the boilerplate doing everything.

Having things in modules starts to enforce a more modular approach and considerations on what to put where.
Will help re-usability. May even be helpful regarding Cmake toolchain file support which has also been requested.

So I would prefer to put it the question the other way around, what exact benefits do you see in having code in `zephyr_default.cmake` that users cannot re-use without having to use that specific module, or from which they cannot opt-out ?

Reply:
This is a misunderstanding on my part. I thought modules were to be user-includable.

Not having something in the cmake/modules directory would have been a way to distinguish between user-includable modules and internal build scripts.

In other words, we would have enforced that users couldn't include internal scripts by not having them on the module include path.

But since (neraly) all modules are internal we don't need to make this distinction.

Reply:
> In other words, we would have enforced that users couldn't include internal scripts by not having them in the module include path.

that's not a bad idea.
The intention is that systems / users who only needs to load certain parts should still go through `find_package(Zephyr COMPONENTS <i-want-this>)`.

And for internal convenience, and to make it clearer which components can be chosen, we have those in a dedicated `cmake/modules` folder, and load them as modules.

But we could remove `cmake/modules` from `CMAKE_MODULE_PATH` when we are done loading modules as part of `find_package()`.
That would make it even clearer that those are part of the package, and not intended to load arbitrary.
(of course users can always do what they want)